### PR TITLE
Ry post page

### PR DESCRIPTION
**Pull request by yamaser**
New Post Feature

App.js, Navbar.js, and router.js updated accordingly.

Updated the sql data file: added UNIQUE constraint to the upc_barcode attribute in the items table to prevent duplicate item entries with the same barcode, updated item_names to be more descriptive based on results from barcodelookup.com, added a new cheese item.

Two new frontend pages are: 1) PostPage 2) PostConfirmationPage

barcodelookup npm package added to package.json.

**barcodelookup API functionality added to router.js.**

NOTE: I signed up for a free account at barcodelookup.com and added my API key to my .env file.

### CODE REVIEW

File: client/src/pages/PostPage.js [hunk 1]
```diff
@@ -1,15 +1,37 @@
...
 
 class PostPage extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      barcode: ''
+    }
+  }
+
+  handleBarcChange = (event) => {
```
Comment (davisethan) [inline client/src/pages/PostPage.js]: Did you fix the type? Should be `handleBarChange`

Comment (davisethan) [inline client/src/pages/PostPage.js]: This button is in a form so you should add a line at the top of your function `event.prevenDefault()`

File: client/src/pages/PostFormPage.js [hunk 2]
```diff
@@ -14,35 +14,73 @@ class PostFormPage extends React.Component {
...
-                            <input id="barcode" type="text" className="form-control" name="barcode" value={this.state.item.barcode} disabled></input>
+                            <input id="barcode" type="text" className="form-control" name="barcode" value={this.state.barcode} disabled></input>
                         </div>
                         <div>
-                            <label htmlFor="tags" className="form-label">Tags comma separated</label>
-                            <input id="tags" type="text" className="form-control" name="tags" value={this.state.tags.join(', ')} onChange={this.handleTagsChange}></input>
+                            <label htmlFor="name" className="form-label">Name</label>
+                            <input id="name" type="text" className="form-control" name="name" value={this.state.item_name} onChange={this.handleNameChange}></input>
+                        </div>
+                        {/* <div>
```
Comment (davisethan) [inline client/src/pages/PostFormPage.js]: Remove commented code from pull requests

File: client/src/pages/PostFormPage.js [hunk 3]
```diff
@@ -14,35 +14,73 @@ class PostFormPage extends React.Component {
...
+                            <label htmlFor="store" className="form-label">Where Are You?</label>
+                            <select id="store" className="form-control" name="store" onChange={this.handleStoreChange}>
+                                <option value="Select a Store"> -- Select a Store -- </option>
+                                {this.state.stores.map((store) => <option value={store.name}>{store.name}</option>)}
+                            </select>
                         </div>
                         <div>
-                            <label htmlFor="quantity" className="form-label">Quantity</label>
-                            <input id="quantity" type="number" className="form-control" name="quantity" value={this.state.quantity} onChange={this.handleQuantityChange}></input>
+                            <label htmlFor="price" className="form-label">Price</label>
```
Comment (davisethan) [inline client/src/pages/PostFormPage.js]: Remove all commented code from pull requests

File: client/src/pages/PostPage.js [hunk 4]
```diff
@@ -11,7 +11,7 @@ class PostPage extends React.Component {
     }
   }
 
-  handleBarcChange = (event) => {
+  handleBarcodeChange = (event) => {
```
Comment (davisethan) [inline client/src/pages/PostPage.js:38]: Ok, you fixed the typo

File: client/src/pages/PostFormPage.js [hunk 5]
```diff
@@ -66,6 +66,7 @@ class PostFormPage extends React.Component {
     handleSubmit = (event) => {
         event.preventDefault();
         console.log(this.state.item.name);
+        console.log(this.state.tags);
```
Comment (davisethan) [inline client/src/pages/PostFormPage.js]: Remove all development console.logs from pull requests

File: server/router.js [hunk 6]
```diff
@@ -72,18 +72,51 @@ router.post('/api/cart/delete', (req, res) => {
...
 
-router.post('/api/post/barcode', (req, res) => {
+router.post('/api/post/create', (req, res) => {
+  const userId = 1; // Passport req.user.userId
   const barcode = req.body.barcode;
+  const name = req.body.name;
+  const selectedStoreId = req.body.selectedStoreId;
+  const price = req.body.price;
+  const onSale = req.body.onSale;
+  console.log(req.body);
```
Comment (davisethan) [inline server/router.js]: Remove all development console.logs from pull requests always

File: server/router.js [hunk 7]
```diff
@@ -72,18 +72,51 @@ router.post('/api/cart/delete', (req, res) => {
...
+      query = `INSERT INTO scans (user_id, item_id, store_id, price, price_time, on_sale) VALUES (?, ?, ?, ?, NOW(), ?)`;
+      db.query(query, [userId, itemId, selectedStoreId, price, onSale], (err, result) => {
+        if (err) throw err;
+
+        // Select new scan id
+        query = `SELECT MAX(scan_id) AS scan_id from scans`;
+        db.query(query, (err, result) => {
+          if (err) throw err;
+          console.log(result);
+
```
Comment (davisethan) [inline server/router.js]: Remove extra blank lines from code for readability

File: client/src/pages/PostPage.js [hunk 8]
```diff
@@ -81,23 +81,23 @@ class PostPage extends React.Component {
         <Navbar />
         <div className="container-fluid">
           <Nav />
-          <form>
+          <form class="needs-validation" onSubmit={this.handleSubmit}>
```
Comment (davisethan) [inline client/src/pages/PostPage.js]: React uses `className` to identify classes in HTML-like syntax. Keyword `class` is reserved to define Javascript classes. Be sure to read your developer tools console for errors before committing code

File: server/router.js [hunk 9]
```diff
@@ -79,6 +79,7 @@ router.post('/api/post/create', (req, res) => {
   const selectedStoreId = req.body.selectedStoreId;
   const price = req.body.price;
   const onSale = req.body.onSale;
+  console.log(name);
```
Comment (davisethan) [inline server/router.js]: Remove developer console.logs before commits

File: server/data/basic_testcase01.sql [hunk 10]
```diff
@@ -102,12 +102,12 @@ INSERT INTO stores (store_name, store_street, store_city, store_state, store_zip
 ('New Seasons Market', '6300 N Lombard St', 'Portland', 'OR', '97203');
 
 INSERT INTO items (item_name, upc_barcode) VALUES
-('toilet paper', '037000940500'),
-('hotdogs', '044700075050'),
-('bread', '072250011372'),
-('milk', '813267020076'),
-('milk', '026400289302'),
-('eggs', '715141729283');
+('Charmin Ultra Soft Toilet Paper, 2-Ply, White, 284 Sheets/Roll, 12 Mega Rolls/Pack', '037000940500'),
```
Comment (davisethan) [inline server/data/basic_testcase01.sql:105]: Thanks for updating the seed data

File: server/router.js [hunk 11]
```diff
@@ -106,20 +106,36 @@ router.post('/api/post/barcode', (req, res) => {
   // Insert new entry in items table only if barcode is new
   let query = `INSERT IGNORE INTO items (item_name, upc_barcode) VALUES ('', ?)`;
   db.query(query, [barcode], (err, result) => {
-    if (result.changedRows == 0) {
+    if (result.affectedRows == 1) {
```
Comment (davisethan) [inline server/router.js:109]: Always use triple equals in Javascript

File: client/src/pages/PostPage.js [hunk 12]
```diff
@@ -76,41 +94,54 @@ class PostPage extends React.Component {
   }
 
   render() {
+    let form;
+    if (!this.state.barcode_sent) {
```
Comment (davisethan) [inline client/src/pages/PostPage.js:98]: Good work finding how to do conditional page renders

**Review by davisethan — COMMENTED** [review]
Good work finding how to use React, Node, and SQL together. You built a full-stack feature congrats! One of my favorite experiences as a developer has been building a project/feature, then doing it again similarly and seeing ways to improve my code design and construction process.

Before merging your pull request, please see my comments.

File: client/src/pages/PostPage.js [hunk 13]
```diff
@@ -23,17 +91,38 @@ class PostPage extends React.Component {
...
+            <button type="submit" className="btn btn-primary" onClick={this.handleSubmit}>Submit</button>
           </form>
         </div>
       </div>
     );
   }
 }
 
-export default PostPage;
+export default WrappedPostPage;
```
[the remark this answers is not in the document]
Comment (yamaser) [reply client/src/pages/PostPage.js]: Great catch. I've addressed this in my latest commit

File: client/src/pages/PostPage.js (hunk 8 above)
Comment (yamaser) [reply client/src/pages/PostPage.js]: Thanks! This has been fixed in my latest commit.

File: server/router.js (hunk 11 above)
Comment (yamaser) [reply server/router.js]: Thanks! This has been addressed in the latest commit.

**Review by davisethan — APPROVED** [review]
Looks ok

**Pull request merged**